### PR TITLE
Bump PHPStan dependency to 0.12

### PR DESCRIPTION
**Pull request by lookyman**
This would of course need a new major version.

### CODE REVIEW

**Comment by ondrejmirtes** [comment]
Great work! :)

> of course need a new major version

I don't think so. Dependency bump is not a BC break. People depending on ^1.0 and on PHPStan 0.11 would continue using version that still supports it.

What I'm missing in this PR is that PHPStan analysis itself could now run on level 8.

**Comment by lookyman** [comment]
> People depending on ^1.0 and on PHPStan 0.11 would continue using version that still supports it.

True.

> What I'm missing in this PR is that PHPStan analysis itself could now run on level 8.

It runs without arguments on Travis, so I think it already is on level 8?

**Comment by ondrejmirtes** [comment]
Nope: https://github.com/Roave/no-floaters/blob/master/phpstan.neon.dist#L11

**Comment by lookyman** [comment]
Now it is 😊

**Comment by ondrejmirtes** [comment]
The ComposerRequireCheck build failure is because PHPStan is now distributed as PHAR. ComposerRequireCheck cannot reason about it. I removed it from PHPStan extensions because of this. (Not sure how hard would it be to fix this issue in it.)

File: src/DisallowFloatAssignedToVariableRule.php [hunk 1]
```diff
@@ -26,10 +30,7 @@ public function getNodeType() : string
         return Node::class;
     }
 
-    /**
-     * @return string[]
-     */
-    public function processNode(Node $node, Scope $scope) : array
+    public function processNode(Node $node, Scope $scope) : array // phpcs:ignore SlevomatCodingStandard.TypeHints.TypeHintDeclaration.MissingTraversableReturnTypeHintSpecification
```
Comment (Ocramius) [inline src/DisallowFloatAssignedToVariableRule.php]: A type declaration for the type being returned, or an `{@inheritDoc}`, are required

File: src/DisallowFloatEverywhereRule.php [hunk 2]
```diff
@@ -8,40 +8,38 @@
...
     public function getNodeType() : string
     {
         return Expr::class;
     }
 
-    /**
-     * @return string[]
-     */
-    public function processNode(Node $node, Scope $scope) : array
+    public function processNode(Node $node, Scope $scope) : array // phpcs:ignore SlevomatCodingStandard.TypeHints.TypeHintDeclaration.MissingTraversableReturnTypeHintSpecification
```
Comment (Ocramius) [inline src/DisallowFloatEverywhereRule.php]: Same as above: type is required here, or at least an `{@inheritDoc}`

File: src/DisallowFloatInFunctionSignatureRule.php [hunk 3]
```diff
@@ -37,13 +41,8 @@ public function getNodeType() : string
         return Function_::class;
     }
 
-    /**
-     * @return string[]
-     */
-    public function processNode(Node $node, Scope $scope) : array
+    public function processNode(Node $node, Scope $scope) : array // phpcs:ignore SlevomatCodingStandard.TypeHints.TypeHintDeclaration.MissingTraversableReturnTypeHintSpecification
```
Comment (Ocramius) [inline src/DisallowFloatInFunctionSignatureRule.php]: Same as above: an `{@inheritDoc}` should suffice

File: src/DisallowFloatInMethodSignatureRule.php [hunk 4]
```diff
@@ -11,30 +11,29 @@
...
     public function getNodeType() : string
     {
         return ClassMethod::class;
     }
 
-    /**
-     * @return string[]
-     */
-    public function processNode(Node $node, Scope $scope) : array
+    public function processNode(Node $node, Scope $scope) : array // phpcs:ignore SlevomatCodingStandard.TypeHints.TypeHintDeclaration.MissingTraversableReturnTypeHintSpecification
```
Comment (Ocramius) [inline src/DisallowFloatInMethodSignatureRule.php]: Same as above: an `{@inheritDoc}` should suffice

File: src/DisallowFloatPropertyTypeRule.php [hunk 5]
```diff
@@ -8,44 +8,42 @@
...
     public function getNodeType() : string
     {
         return PropertyProperty::class;
     }
 
-    /**
-     * @return string[]
-     */
-    public function processNode(Node $node, Scope $scope) : array
+    public function processNode(Node $node, Scope $scope) : array // phpcs:ignore SlevomatCodingStandard.TypeHints.TypeHintDeclaration.MissingTraversableReturnTypeHintSpecification
```
Comment (Ocramius) [inline src/DisallowFloatPropertyTypeRule.php]: Same as above: an `{@inheritDoc}` should suffice

**Review by Ocramius — CHANGES_REQUESTED** [review]

**Comment by Ocramius** [comment]
> ComposerRequireCheck cannot reason about it.

How are classes autoloaded in the new package? Just loading the entire phar?

**Comment by ondrejmirtes** [comment]
If you're only using PHPStan-referencing code when running PHPStan, the autoloader from inside the PHAR is used.

If you're referencing PHPStan code when not running PHPStan, so for example for the rules unit tests, then this gets executed: https://github.com/phpstan/phpstan/blob/master/bootstrap.php#L29-L36

When some prefixed dependency gets required when not running PHPStan, then this is executed: https://github.com/phpstan/phpstan/blob/master/bootstrap.php#L9-L16 (this will also most likely happen when running RuleTestCase because part of PHPStan gets booted and used)

**Comment by Ocramius** [comment]
Yeah, I think that's impossible to check for `ComposerRequireChecker` for now, unless we completely switch to BetterReflection there. Even then, the `ComposerRequireAutoloader` is deprecated, as it is unsafe to load third party scripts to perform autoloading of unknown sources.

I think it is safe to drop it from the CI pipeline and dependencies here, @lookyman.

**Comment by lookyman** [comment]
@Ocramius Done.

**Comment by Ocramius** [comment]
Ouff, `roave/backwards-compatibility-check` also cannot find these classes :-( This entire phar business is a massive headache...

I'll think about a solution for that

**Comment by ondrejmirtes** [comment]
Why? They can be autoloaded by Composer itself...

**Comment by Ocramius** [comment]
These are not in deterministic locations, declared via a composer package definition, but rather through a custom autoloader.

I understand the reason why this is done, it just isn't something usable by tooling, which is understandable too.

**Comment by fezfez** [comment]
Any news ?

**Comment by Ocramius** [comment]
the BC check won't work at all right now, sorry - can't merge in current state.

Either we completely drop it (this is only infrastructure implementing upstream types, after all), or we find a way to make this work with PHARs (rabbit hole)

**Comment by ondrejmirtes** [comment]
I'm for merging this and tagging the next minor release. We can solve the infrastructure another time, it's apparent this code works and people would benefit from it.

**Comment by Ocramius** [comment]
If @lookyman can remove the check from CI, then we should be OK to merge then.

I think PHAR introspection in better-reflection may never actually happen, unless someone wants to maintain it.

**Comment by lookyman** [comment]
Ok, I'll do that later today.

**Review by Ocramius — APPROVED** [review]
LGTM :+1:

**Comment by Ocramius** [comment]
And released: https://github.com/Roave/no-floaters/releases/tag/1.3.0

**Comment by fezfez** [comment]
Thanks all!

**Pull request merged**